### PR TITLE
[WIP] - Testing new nav idea

### PR DESCRIPTION
**Pull request by psav**
You can do class or instance based navigation with the same method,

```python
from cfme.cloud.provider import Provider
Provider.navigate('New')
```

or

```python
from utils.providers import get_crud
prov = get_crud('rhos7-ga')
prov.navigate('Details')
```

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 91f1c20821a2d87302d7e22fe62102528cb99763:

`cfme/cloud/provider/__init__.py`:
- :red_circle: Line 169:7: **N801** *class names should use CapWords convention*
- :red_circle: Line 180:7: **N801** *class names should use CapWords convention*
- :red_circle: Line 189:7: **N801** *class names should use CapWords convention*

`utils/appliance/__init__.py`:
- :red_circle: Line 1634:7: **N801** *class names should use CapWords convention*
- :red_circle: Line 1643:7: **N801** *class names should use CapWords convention*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 0c8e905398dabbcf4b7f4e97f52ce7b809a3c2a3:

`cfme/cloud/provider/__init__.py`:
- :red_circle: Line 169:7: **N801** *class names should use CapWords convention*
- :red_circle: Line 180:7: **N801** *class names should use CapWords convention*
- :red_circle: Line 189:7: **N801** *class names should use CapWords convention*

`utils/appliance/__init__.py`:
- :red_circle: Line 1634:7: **N801** *class names should use CapWords convention*
- :red_circle: Line 1643:7: **N801** *class names should use CapWords convention*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by jkrocil** [comment]
I like it but I'd like to see it called `navigate` both for classes and instances so that the following works:
```
Provider.navigate("new")
prov.navigate("new")         # not sure why someone would do that
prov.navigate("details")
```

Also, the navigation classes should probably not be defined in module scope if there's gonna be classes with names like `all`, `list` and similar with potential to override python keywords.

**Comment by psav** [comment]
Yeh, unfortunately I don't think we can actually have them both named the same, I really wanted that too, but perhaps it could serve as a reminder to people. You can actually call the class based method with instance_navigate, but you can't call an instance method with navigate, so perhaps we should rename them, ```navigate``` and ```class_navigate```?

**Comment by dajoRH** [comment]
Lint report for commit 87140a2f0244b486254e3b6bcd644b4b289b30a0:

`utils/navigate.py`:
- :red_circle: Line 1:7: **N801** *class names should use CapWords convention*
- :grey_exclamation: Line 25:9: **T001** *print statement found.*
- :grey_exclamation: Line 35:13: **T001** *print statement found.*
- :grey_exclamation: Line 42:13: **T001** *print statement found.*
- :grey_exclamation: Line 44:9: **T001** *print statement found.*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 5dc5400dc49026fc6d479339c83f2bf1f8074533:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by psav** [comment]
So I now implemented the super monkey patch, I think it's justified here

**Comment by dajoRH** [comment]
Lint report for commit a984ba211c4aceca4cb96148ceaabaa7f3b0b224:

`utils/navigate.py`:
- :red_circle: Line 24:9: **N805** *first argument of a method should be named 'self'*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2293dd9a0329e27db3e8355a3675dd78313bb283:

`utils/navigate.py`:
- :red_circle: Line 24:9: **N805** *first argument of a method should be named 'self'*

Please, rectify these issues :smirk: .
*CFME QE Bot*

File: utils/navigate.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    def __init__(self, *args, **kwargs):
+        self.navigate = self.instance_navigate
+
+    @InstanceClassMethod
+    def navigate(self, name):
+        nav_step = self._nav_steps[name](self)
+        nav_step.go()
+
+    @navigate.classmethod
+    def navigate(cls, name):
```
Comment (mfalesni) [inline utils/navigate.py]:
```python
def navigate(cls, name):  # noqa
```

Should shut it up :)
flake is trying to be smarter than it is necessary :)

**Comment by dajoRH** [comment]
Lint report for commit e5751cada53bd402c1d80942fb726385cf88ed0c:

`utils/navigate.py`:
- :red_circle: Line 24:9: **N805** *first argument of a method should be named 'self'*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8d7a166375f9cd22d5b776431c55b7d89157b595:

Everything seems all right :smile: .
*CFME QE Bot*

File: utils/navigate.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        nav_step = self._nav_steps[name](self)
+        nav_step.go()
+
+    @navigate.classmethod
+    def navigate(cls, name):  # noqa
+        nav_step = cls._nav_steps[name](cls)
+        nav_step.go()
+
+    @classmethod
+    def navigatable(cls, *args, **kwargs):
```
Comment (mfalesni) [inline utils/navigate.py]: What will this one be good for?

Comment (psav) [inline utils/navigate.py]: I think is was left over........

Comment (psav) [inline utils/navigate.py]: I can remove

File: utils/navigate.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        return args[0]
+
+
+class NavigateStep(object):
+    def __init__(self, obj):
+        self.obj = obj
+
+    def go(self):
+        if hasattr(self, 'am_i_here'):
```
Comment (RonnyPfannschmidt) [inline utils/navigate.py]: we should just have a default implementation returning False

Comment (psav) [inline utils/navigate.py]: Noted - good idea

**Comment by dajoRH** [comment]
Lint report for commit c241b3156bb9363193021f8dc94e5222e4fb33de:

Everything seems all right :smile: .
*CFME QE Bot*

File: utils/navigate.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        self.cls._nav_steps[f.__name__] = f
+        return f
+
+
+class Navigatable(object):
+    def __init__(self, *args, **kwargs):
+        self.navigate = self.instance_navigate
+
+    @InstanceClassMethod
+    def navigate(self, name):
```
Comment (RonnyPfannschmidt) [inline utils/navigate.py]: i think for clarity we might want to replace this one iwth a `ClassORInstanceMethod` that has the prototype `def navigate(cls_or_self, name):`  ´</nitpick>`

File: utils/navigate.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+from utils import InstanceClassMethod
+from utils.log import logger
+
+
+class NavigateWrap(object):
```
Comment (RonnyPfannschmidt) [inline utils/navigate.py]: can we reduce this one to a parameterized decorator function?

File: utils/navigate.py (hunk 4 above)
Comment (mfalesni) [inline utils/navigate.py]: I am not that big fan of this approach, though I guess the class could get a param that will set that one func can get invoked in both contexts ...

**Comment by dajoRH** [comment]
Lint report for commit 4cab9ac435f16e1a3da23efb5101b86e21a4fcb9:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit d8f907590f87f18fbae824881f647ce6fc9e2d40:

Everything seems all right :smile: .
*CFME QE Bot*

**Pull request closed without merging**